### DISCUSSION TITLE
Acquiring camera image from supervisor

### DISCUSSION THREAD

**Discussion by janblumenkamp** [cyberbotics/webots#4524] [Q&A]
I am trying to implement a controller that runs a control policy for multiple devices/robots. Each device has a camera, of which I would like to acquire the camera image. This should be possible according to [this statement](https://cyberbotics.com/doc/guide/tutorial-8-the-supervisor?tab-language=python#acquire-measurements-using-a-supervisor), since the supervisor has unlimited power. The documentation is not super clear how exactly I can retrieve information from a specific device through the supervisor though. How can I retrieve individual nodes' camera images from the supervisor?

**Comment by stefaniapedrazzi** [comment]
It is not possible to retrieve internal device information or use the devices API from controller associated to a different Robot node, including a [Supervisor](https://www.cyberbotics.com/doc/reference/supervisor).
The "unlimited power" of a [Supervisor](https://www.cyberbotics.com/doc/reference/supervisor) is basically that can access all the information displayed in the scene tree but the camera image or any other device measurement is not displayed in the scene tree and cannot be accessed.
If I would make an analogy with a real experiment, a [Supervisor](https://www.cyberbotics.com/doc/reference/supervisor) is like you that can interact with the environment and have all the information about the environment. But also in a real environment you don't have access to the robot device measurements out-of-the-box, instead you need some code to retrieve them from the device.

So, to solve your issue you should first make sure that you really need to share the camera images between different controller processes, for example:
* are the devices/robots moving independently and do you really need to model them as independent Robot nodes?
* is it possible to pre-process the images in the robot controller and only send results to the Supervisor controller?

Then, if there are no other solutions, to share the device measurements among different controllers you have to use some inter-process communication methods (i.e. sockets, storing in a file, shared memory, etc.).
Note that Webots also provides the [Emitter](https://www.cyberbotics.com/doc/reference/emitter) and [Receiver](https://www.cyberbotics.com/doc/reference/receiver) nodes that allows to share information between controller programs, but it isn't the best option to share images.

**Reply by janblumenkamp** [reply]
That makes sense, thanks for clarifying!
